database/buildSchema/43_views_functions_triggers.sql

```diff
@@ -1542,164 +1542,381 @@ $$
+        LEFT JOIN review AS action_review ON action_review.review_assignment_id = reviewer_assignment.id
+        LEFT JOIN review_status_history AS action_review_status_history ON (action_review_status_history.review_id = action_review.id
+                AND action_review_status_history.is_current = TRUE)
+            -- ASSIGNER ACTIONS
+        LEFT JOIN assigned_sections_by_stage_and_level ON (assigned_sections_by_stage_and_level.application_id = app.id
+                AND assigned_sections_by_stage_and_level.stage_id = stage_status.stage_id
+                AND app.outcome = 'PENDING'
+                AND assigned_sections_by_stage_and_level.assigner_id = userId)
+        LEFT JOIN template_section AS assignable_sections ON assignable_sections.template_id = template.id
+        LEFT JOIN allowed_self_assignable_sections(userId) AS allowed_self_assignable_sections ON allowed_self_assignable_sections.review_assignment_id = reviewer_assignment.id
+        LEFT JOIN review_assignment_assigned_sections(userId) as review_assignment_assigned_sections ON review_assignment_assigned_sections.review_assignment_id = reviewer_assignment.id
+    WHERE
+        app.id = applicationId
+    GROUP BY stage_status.stage_id;
+$$
+LANGUAGE sql
+STABLE;
+
+-- Function to update assigner/reviewer lists on applications and insert/update
+-- reviewer/assigner actions on application_reviewer_action, by making calls to
+-- the above `single_application_detail` function for the appropriate
+-- user/application combo(s)
+DROP FUNCTION IF EXISTS public.update_application_reviewer_stats CASCADE;
+CREATE OR REPLACE FUNCTION public.update_application_reviewer_stats ()
+    RETURNS TRIGGER
+     -- This allows the function to be run as admin user, otherwise it'll only
+     -- be able to query rows the current user has access to.
+    SECURITY DEFINER
+    AS $$
+
+DECLARE
+    reviewer_ids integer[];
+    rev_id integer;
+    assigner_ids integer[];
+    ass_id integer;
+
+BEGIN
+-- reviewer and assigner lists
+    IF TG_OP != 'INSERT' THEN
+        WITH lists AS (
+                SELECT reviewers, assigners
+                FROM single_application_detail(NEW.application_id)
+            ) 
+        UPDATE public.application
+            SET reviewer_list = (SELECT reviewers FROM lists),
+            assigner_list = (SELECT assigners FROM lists)
+            WHERE id = NEW.application_id;
+    END IF;
+
+    -- Get list of reviewer_ids to update for 
+    IF TG_OP = 'UPDATE' THEN
+        reviewer_ids = ARRAY(
+            SELECT reviewer_id FROM review_assignment
+            WHERE application_id = NEW.application_id
+        );
+    ELSE
+        reviewer_ids = ARRAY[NEW.reviewer_id];
+    END IF;
+
+    FOREACH rev_id IN ARRAY reviewer_ids LOOP
+        -- Remove existing
+        UPDATE public.application_reviewer_action
+            SET reviewer_action = NULL
+            WHERE application_id = NEW.application_id
+            AND user_id = rev_id;
```

Comment:
I tried a few different approaches for handling the existing records, ended up with:
- Set existing value to `NULL` (the reason it's not deleted is that it processes the reviewer action separately to the assigner action, so one of them might not be getting changed)
- Deduce reviewer_action, update record
- Deduce assigner_action, update record
- Once complete, if any records end up with `NULL` values in both fields, delete them (to keep the list from growing forever and filling up with obsolete records)